File: 1300-1399/1358.Number-of-Substrings-Containing-All-Three-Characters/Solution.java
# Java Solution

```java
class Solution {
    public int numberOfSubstrings(String s) {
        int[] count = new int[3];
        int left = 0;
        int answer = 0;
        int n = s.length();

        for (int right = 0; right < n; right++) {
            count[s.charAt(right) - 'a']++;

            while (count[0] > 0 && count[1] > 0 && count[2] > 0) {
                answer += n - right;
                count[s.charAt(left) - 'a']--;
                left++;
            }
        }

        return answer;
    }
}

/*

Alternative Approach (Last Seen Indices)

// T.C : O(n)
// S.C : O(1)

class Solution {
    public int numberOfSubstrings(String s) {

        int[] lastSeen = {-1, -1, -1};

        int answer = 0;

        for (int i = 0; i < s.length(); i++) {

            lastSeen[s.charAt(i) - 'a'] = i;

            int minIndex = Math.min(
                lastSeen[0],
                Math.min(lastSeen[1], lastSeen[2])
            );

            if (minIndex != -1) {

                answer += minIndex + 1;
            }
        }

        return answer;
    }
}

*/
```

File: 1300-1399/1358.Number-of-Substrings-Containing-All-Three-Characters/Solution.cpp
#C++ Solution

```cpp class Solution
{
public:
  int numberOfSubstrings(string s)
  {

    vector<int> count(3, 0);

    int left = 0;

    int answer = 0;

    int n = s.size();

    for (int right = 0; right < n; right++)
    {

      count[s[right] - 'a']++;

      while (count[0] > 0 &&
             count[1] > 0 &&
             count[2] > 0)
      {

        answer += n - right;

        count[s[left] - 'a']--;

        left++;
      }
    }

    return answer;
  }
};

/*

Alternative Approach (Last Seen Indices)

// T.C : O(n)
// S.C : O(1)

class Solution {
public:
    int numberOfSubstrings(string s) {

        vector<int> lastSeen(3, -1);

        int answer = 0;

        for (int i = 0; i < s.size(); i++) {

            lastSeen[s[i] - 'a'] = i;

            int minIndex = min(
                lastSeen[0],
                min(lastSeen[1], lastSeen[2])
            );

            if (minIndex != -1) {

                answer += minIndex + 1;
            }
        }

        return answer;
    }
};

*/
